cpp/kiss_icp/core/VoxelHashMap.cpp

```diff
@@ -56,6 +56,17 @@ std::vector<Eigen::Vector3d> VoxelHashMap::Pointcloud() const {
     return points;
 }
 
+// Get the indices of the occupided voxels as points, mainly used for visualization
+std::vector<Voxel> VoxelHashMap::GetVoxels() const {
```

Comment:
I would write this function with a transform, just to be pedantic :) 
```suggestion
std::vector<Voxel> VoxelHashMap::GetVoxels() const {
std::vector<Voxel> voxels(map_.size());
    std::transform(map_.cbegin(), map_.cend(), voxels.begin(),
                   [](const auto &map_element) { return map_element.first; });
    return voxels;
}
```

Reply:
I never liked this `transform` operations as it confuses me when I read it, we are not transforming anything, we are just extracting data. And 2, it's typically slower, as the `std::vector<Voxel> voxels(map_.size())` will construct `map.size()` voxels calling the default constructor and later on the `std::transfrom` will iterate over all the constructed elements, modifying those in place.

If you have a STL algorithm that sucks the voxels out of the hash map then I take it :nerd_face:

Reply:
That's partially not true, my friend; I'm guilty as I was the person promoting this idea at the lab :P . If we check the STL implementation of `std::transform`
```cpp
  /**
   *  @brief Perform an operation on a sequence.
   *  @ingroup mutating_algorithms
   *  @param  __first     An input iterator.
   *  @param  __last      An input iterator.
   *  @param  __result    An output iterator.
   *  @param  __unary_op  A unary operator.
   *  @return   An output iterator equal to @p __result+(__last-__first).
   *
   *  Applies the operator to each element in the input range and assigns
   *  the results to successive elements of the output sequence.
   *  Evaluates @p *(__result+N)=unary_op(*(__first+N)) for each @c N in the
   *  range @p [0,__last-__first).
   *
   *  @p unary_op must not alter its argument.
  */
  template<typename _InputIterator, typename _OutputIterator,
	   typename _UnaryOperation>
    _GLIBCXX20_CONSTEXPR
    _OutputIterator
    transform(_InputIterator __first, _InputIterator __last,
	      _OutputIterator __result, _UnaryOperation __unary_op)
    {
      for (; __first != __last; ++__first, (void)++__result)
	      *__result = __unary_op(*__first);
      return __result;
    }
```

What happens is that sometimes using STL, besides being more precise to express what you want to do, does lead to optimization when compared with our hand-baked shitty code. In this particular case, `const auto &[voxel, voxel_block] : map_` vs. `map_element.first` might be slower, and the STL algorithm forces us to write this in a "better way." But is not doing much more magic in this specific scenario ;)

Reply:
By "optimization" I meant: once compiled probably "reserving and emplacing_back" vs. "pre-constructing and modifing in place" is the same in terms of binary code (but I'm guessing here, no idea how compilation optimizes the two things at the end, probably better from my side to actually check this before guessing and talking). Aware that no magic happens with STL algorithms. In this silly scenario I think is more a matter of taste that practical difference, that was my claim :) And the idea of "trasforming" a voxel_map into a vector of voxel through a `std::transform` was juicy :yum:

Reply:
> And the idea of "trasforming" a voxel_map into a vector of voxel through a std::transform was juicy ... 

 start like this, now you tricked my brain